syntax = "proto3";

option java_package = "com.datadoghq.client";
//option go_package = "./";
option csharp_namespace = "ApmTestClient";

// Interface of APM clients to be used for shared testing.
service APMClient {
  rpc StartSpan(StartSpanArgs) returns (StartSpanReturn) {}
  rpc FinishSpan(FinishSpanArgs) returns (FinishSpanReturn) {}
  rpc SpanName(SpanNameArgs) returns (SpanNameReturn) {}
  rpc SpanResource(SpanResourceArgs) returns (SpanResourceReturn) {}
  rpc SpanMeta(SpanMetaArgs) returns (SpanMetaReturn) {}
  rpc SpanMetric(SpanMetricArgs) returns (SpanMetricReturn) {}
  rpc SpanSetMeta(SpanSetMetaArgs) returns (SpanSetMetaReturn) {}
  rpc SpanSetMetric(SpanSetMetricArgs) returns (SpanSetMetricReturn) {}
  rpc SpanSetError(SpanSetErrorArgs) returns (SpanSetErrorReturn) {}
  rpc SpanSetResource(SpanSetResourceArgs) returns (SpanSetResourceReturn) {}
  rpc SpanAddLink(SpanAddLinkArgs) returns (SpanAddLinkReturn) {}
  rpc HTTPClientRequest(HTTPRequestArgs) returns (HTTPRequestReturn) {}
  rpc HTTPServerRequest(HTTPRequestArgs) returns (HTTPRequestReturn) {}
  rpc InjectHeaders(InjectHeadersArgs) returns (InjectHeadersReturn) {}
  rpc FlushSpans(FlushSpansArgs) returns (FlushSpansReturn) {}
  rpc FlushTraceStats(FlushTraceStatsArgs) returns (FlushTraceStatsReturn) {}
  rpc CurrentSpan(CurrentSpanArgs) returns (CurrentSpanReturn) {}

  rpc OtelStartSpan(OtelStartSpanArgs) returns (OtelStartSpanReturn) {}
  rpc OtelEndSpan(OtelEndSpanArgs) returns (OtelEndSpanReturn) {}
  rpc OtelIsRecording(OtelIsRecordingArgs) returns (OtelIsRecordingReturn) {}
  rpc OtelSpanContext(OtelSpanContextArgs) returns (OtelSpanContextReturn) {}
  rpc OtelSetStatus(OtelSetStatusArgs) returns (OtelSetStatusReturn) {}
  rpc OtelSetName(OtelSetNameArgs) returns (OtelSetNameReturn) {}
  rpc OtelSetAttributes(OtelSetAttributesArgs) returns (OtelSetAttributesReturn) {}
  rpc OtelFlushSpans(OtelFlushSpansArgs) returns (OtelFlushSpansReturn) {}
  rpc OtelFlushTraceStats(OtelFlushTraceStatsArgs) returns (OtelFlushTraceStatsReturn) {}
  rpc OtelCurrentSpan(OtelCurrentSpanArgs) returns (OtelCurrentSpanReturn) {}
  rpc OtelGetAttribute(OtelGetAttributeArgs) returns (OtelGetAttributeReturn) {}
  rpc OtelGetName(OtelGetNameArgs) returns (OtelGetNameReturn) {}
  rpc OtelGetLinks(OtelGetLinksArgs) returns (OtelGetLinksReturn) {}

  rpc StopTracer(StopTracerArgs) returns (StopTracerReturn) {}
}

message StartSpanArgs {
  string name = 1;
  optional string service = 2;
  optional uint64 parent_id = 3;
  optional string resource = 4;
  optional string type = 5;
  optional string origin = 6;
  optional DistributedHTTPHeaders http_headers = 7;
  repeated HeaderTuple span_tags = 8;
  repeated SpanLink span_links = 9;
}

message DistributedHTTPHeaders {
  repeated HeaderTuple http_headers = 1;
}

message SpanLink {
  oneof from {
    uint64 parent_id = 1;
    DistributedHTTPHeaders http_headers = 2;
    uint64 trace_id = 4;
    string tracestate = 5;
  }
  Attributes attributes = 3;
}

message HeaderTuple {
  string key = 1;
  string value = 2;
}

message StartSpanReturn {
  uint64 span_id = 1;
  uint64 trace_id = 2;
}

message InjectHeadersArgs {
  uint64 span_id = 1;
}

message InjectHeadersReturn {
  optional DistributedHTTPHeaders http_headers = 1;
}

message FinishSpanArgs {
  uint64 id = 1;
}
message FinishSpanReturn {}

message SpanNameArgs {
  uint64 span_id = 1;
}
message SpanNameReturn {
  string name = 1;
}

message SpanResourceArgs {
  uint64 span_id = 1;
}
message SpanResourceReturn {
  string resource = 1;
}

message SpanMetaArgs {
  uint64 span_id = 1;
  string key = 2;
}
message SpanMetaReturn {
  string value = 1;
}

message SpanMetricArgs {
  uint64 span_id = 1;
  string key = 2;
}
message SpanMetricReturn {
  float value = 1;
}

message SpanSetMetaArgs {
  uint64 span_id = 1;
  string key = 2;
  string value = 3;
}
message SpanSetMetaReturn {}


message SpanSetMetricArgs {
  uint64 span_id = 1;
  string key = 2;
  float value = 3;
}
message SpanSetMetricReturn {}


message SpanSetErrorArgs {
  uint64 span_id = 1;
  optional string type = 2;
  optional string message = 3;
  optional string stack = 4;
}
message SpanSetErrorReturn {}

message SpanSetResourceArgs {
  uint64 span_id = 1;
  string resource = 2;
}
message SpanSetResourceReturn {}

message SpanAddLinkArgs {
  uint64 span_id = 1;
  repeated SpanLink links = 2;
}

message SpanAddLinkReturn {}


message HTTPRequestArgs {
  string url = 1;
  string method = 2;
  DistributedHTTPHeaders headers = 3;
  bytes body = 4;
}
message HTTPRequestReturn {
  string status_code = 1;
}

message FlushSpansArgs {}
message FlushSpansReturn {}

message FlushTraceStatsArgs {}
message FlushTraceStatsReturn {}

message CurrentSpanArgs {}
message CurrentSpanReturn {
  uint64 span_id = 1;
  uint64 trace_id = 2;
  uint64 parent_id = 3;
}


// OTEL MESSAGES

message OtelStartSpanArgs {
  string name = 1;
  optional uint64 parent_id = 3;
  optional uint64 span_kind = 9;
  optional string service = 4;
  optional string resource = 5;
  optional string type = 6;
  optional int64 timestamp = 7;
  repeated SpanLink span_links = 11;
  optional DistributedHTTPHeaders http_headers = 10;
  Attributes attributes = 8;
}

message OtelStartSpanReturn {
  uint64 span_id = 1;
  uint64 trace_id = 2;
}
message OtelEndSpanArgs {
  uint64 id = 1;
  optional int64 timestamp = 2;
}
message OtelEndSpanReturn {}

message OtelForceFlushArgs{
  uint32 seconds = 1;
}
message OtelForceFlushReturn{
  bool success = 1;
}

message OtelFlushSpansArgs {
  uint32 seconds = 1;
}
message OtelFlushSpansReturn {
  bool success = 1;
}

message OtelFlushTraceStatsArgs {}
message OtelFlushTraceStatsReturn {}

message OtelStopTracerArgs {}
message OtelStopTracerReturn {}

message OtelIsRecordingArgs {
  uint64 span_id = 1;
}
message OtelIsRecordingReturn {
  bool is_recording = 1;
}

message OtelSpanContextArgs {
  uint64 span_id = 1;
}
message OtelSpanContextReturn {
  string span_id = 1;
  string trace_id = 2;
  string trace_flags = 3;
  string trace_state = 4;
  bool remote = 5;
}

message OtelSetStatusArgs {
  uint64 span_id = 1;
  string code = 2;
  string description = 3;
}
message OtelSetStatusReturn {}

message OtelSetNameArgs {
  uint64 span_id = 1;
  string name = 2;
}
message OtelSetNameReturn {}

message OtelSetAttributesArgs {
  uint64 span_id = 1;
  Attributes attributes = 2;
}
message OtelSetAttributesReturn {}

message OtelCurrentSpanArgs {}
message OtelCurrentSpanReturn {
  uint64 span_id = 1;
  uint64 trace_id = 2;
  uint64 parent_id = 3;
}

message OtelGetAttributeArgs {
  uint64 span_id = 1;
  string key = 2;
}
message OtelGetAttributeReturn {
  AttrVal value = 1;
}

message OtelGetNameArgs {
  uint64 span_id = 1;
}
message OtelGetNameReturn {
  string name = 1;
}

message OtelGetLinksArgs {
  uint64 span_id = 1;
}
message OtelGetLinksReturn {
  repeated SpanLink links = 1;
}

message Attributes {
  map<string, ListVal> key_vals = 3;
}

message ListVal {
  repeated AttrVal val = 1;
}

message AttrVal {
  oneof val {
    bool bool_val = 1;
    string string_val = 2;
    double double_val = 3;
    int64 integer_val = 4;
  }
}

message StopTracerArgs {}
message StopTracerReturn {}
